storage/namespace_readers.go

```diff
@@ -0,0 +1,313 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
```

Comment:
Gettin' dangerously close to implementing sessions here ;)

A file-level or struct-level comment explaining what a databaseNamespaceReaderManager is and why its important would be great. Also might be nicer to rename to it databaseNamespaceReaderCache because primary purpose seems to be caching not lifecycle management (it just happens to do that as a side-effect of caching)

> databaseNamespaceReaderManager maintains a pool of closed readers which can be re-used (to prevent additional allocations), as well as a cache of recently used open readers based on their position. The cache of recently used open readers is useful during peer bootstrapping because a pageToken (which contains an offset into the reader for both the data and metadata portions of the fileset) is used to communicate the clients current position to the server. In the general case, the client will miss on its first request for a given shard/block start, and then experience a cache hit on every subsequent request because the current client implementation does not perform any parallel requests within a given shard.
> 
> The closedReaders pool is modeled as a stack (implemented via slice operations) and the open readers cache is implemented as a map where the key is of type cachedOpenReaderKey.
> 
> the databaseNamespaceReaderManager also implements a tick() method which should be called regularly in order to shrunk the closedReaders stack after bursts of usage, as well as to expire cached open readers which have not been used for a configurable number of ticks.

Reply:
Haha yes, this was almost a sessions implementation.  Thankfully if a cache miss occurs the caller doesn't have to start all the way from the start again however.

Thanks for the writeup, very useful.  I'll add it to the file.  I mainly skip these in the interest of time, but we should definitely be adding them.